### PR TITLE
Enumerable project

### PR DESCRIPTION
**Pull request by Jarfsoft**
## Enumerable project

**Recreated enumerable methods**

- each
- each with index
- select
- all?
- any?
- none?
- count
- inject

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -1 +1,122 @@
...
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    count = 0
+    while count < to_a.length
+      yield to_a[count]
+      count += 1
+    end
+  end
```
Comment (maelfosso) [inline main.rb]:
```rb
2.6.5 :026 > block = proc { |num| num < 4 }
2.6.5 :027 > array
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :028 > array.my_each(&block)
 => nil 
2.6.5 :029 > array.each(&block)
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :030 > 
```

Comment (maelfosso) [reply main.rb]:
```rb
2.6.5 :032 > range
 => 5..50 
2.6.5 :033 > block = proc { |num| num < 4 }
2.6.5 :034 > range.each(&block)
 => 5..50 
2.6.5 :035 > range.my_each(&block)
 => nil 
2.6.5 :036 > 
```

File: main.rb [hunk 2]
```diff
@@ -1 +1,122 @@
...
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    count = 0
+    while count < to_a.length
+      yield(to_a[count], count)
+      count += 1
+    end
+  end
```
Comment (maelfosso) [inline main.rb]:
```rb
2.6.5 :038 > block = proc { |num| num < 4 }
2.6.5 :039 > array
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :040 > array.my_each_with_index(&block)
 => nil 
2.6.5 :041 > array.each_with_index(&block)
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :042 > range
 => 5..50 
2.6.5 :043 > range.my_each_with_index(&block)
 => nil 
2.6.5 :044 > range.each_with_index(&block)
 => 5..50 
2.6.5 :045 >
```

File: main.rb [hunk 3]
```diff
@@ -1 +1,122 @@
...
+    if block_given?
+      to_a.my_each { |i| return false if yield(i) == false }
+      return true
+    elsif value.nil?
+      to_a.my_each { |i| return false if i.nil? || i == false }
+    else
+      to_a.my_each { |i| return false if i != value }
+    end
+    true
+  end
```
Comment (maelfosso) [inline main.rb]:
```rb
2.6.5 :047 > array
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :048 > array.my_all?(Integer)
 => false 
2.6.5 :049 > array.all?(Integer)
 => true 
2.6.5 :050 > 
```

Comment (maelfosso) [reply main.rb]:
```rb
2.6.5 :052 > numbers
 => [1, (0+2i), 3.14] 
2.6.5 :053 > numbers.my_all?(Numeric)
 => false 
2.6.5 :054 > numbers.all?(Numeric)
 => true 
2.6.5 :055 > 
```

Comment (maelfosso) [reply main.rb]:
```rb
2.6.5 :057 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :058 > words.my_all?(/d/)
 => false 
2.6.5 :059 > words.all?(/d/)
 => true 
2.6.5 :060 >
```

File: main.rb [hunk 4]
```diff
@@ -1 +1,122 @@
...
+    if block_given?
+      to_a.my_each { |i| return false if yield(i) == false && i.nil? }
+      return true
+    elsif to_a.my_each { |i| return false if i == false }
+      return true
+    else
+      to_a.my_each { |i| return true if i == param }
+    end
+    false
+  end
```
Comment (maelfosso) [inline main.rb]:
```rb
2.6.5 :062 > false_block = proc { |num| num > 9 }
2.6.5 :063 > array
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :064 > array.my_any?(&false_block)
 => true 
2.6.5 :065 > array.any?(&false_block)
 => false 
2.6.5 :066 >
```

Comment (maelfosso) [reply main.rb]:
```rb
2.6.5 :068 > true_array = [nil, false, true, []]
2.6.5 :069 > true_array.any?
 => true 
2.6.5 :070 > true_array.my_any?
 => false 
2.6.5 :071 > 
```

Comment (maelfosso) [reply main.rb]:
```rb
2.6.5 :073 > array
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :074 > array.my_any?(Numeric)
 => false 
2.6.5 :075 > array.any?(Numeric)
 => true 
2.6.5 :076 > 
```

Comment (maelfosso) [reply main.rb]:
```rb
2.6.5 :078 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :079 > words.my_any?(/d/)
 => false 
2.6.5 :080 > words.any?(/d/)
 => true 
2.6.5 :081 >
```

File: main.rb [hunk 5]
```diff
@@ -1 +1,122 @@
...
+
+  def my_none?(value = nil, &block)
+    !to_a.my_any?(value, &block)
+  end
```
Comment (maelfosso) [inline main.rb:79]: It depends on `#my_any`. 
Kindly fix the `my_any` bugs

File: main.rb [hunk 6]
```diff
@@ -1 +1,122 @@
...
+
+  def my_count?(value = nil)
```
Comment (maelfosso) [inline main.rb]: It's `my_count` instead of `my_count?`

File: main.rb [hunk 7]
```diff
@@ -1 +1,122 @@
...
+                  @result += storage
+                elsif !block_given? && param2.nil?
+                  @result.send(param1, storage)
+                else
+                  @result = yield(@result, storage)
+                end
+    end
+    @result
+  end
+end
```
Comment (maelfosso) [inline main.rb]:
```rb
2.6.5 :083 > array
 => [1, 0, 8, 4, 6, 8, 7, 2, 1, 7, 4, 0, 3, 0, 5, 2, 1, 2, 0, 0, 8, 8, 3, 3, 1, 0, 5, 7, 7, 7, 2, 1, 0, 4, 4, 7, 2, 5, 2, 6, 4, 0, 2, 7, 7, 2, 2, 8, 8, 6, 2, 4, 0, 6, 8, 0, 8, 7, 2, 0, 4, 8, 2, 1, 3, 3, 2, 0, 8, 3, 5, 1, 8, 7, 4, 7, 6, 4, 6, 6, 3, 8, 3, 5, 1, 0, 5, 8, 8, 0, 5, 4, 5, 5, 4, 5, 0, 1, 4, 2] 
2.6.5 :084 > array.my_inject
 => 390 
2.6.5 :085 > array.inject
Traceback (most recent call last):
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        3: from (irb):85
        2: from (irb):85:in `inject'
        1: from (irb):85:in `each'
LocalJumpError (no block given)
2.6.5 :086 >
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle:

Hello,

Thanks for the well-done job so far
However, there are still some issues

1- There are some linters' errors. Fix them, please.

2- Read the comments left below to know more about the bugs.
Fix them and submit a new code review request

Thanks
Mael FOSSO

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Marcelo Araújo & Juan Raudales Flores** 
Good job with the improvements for this project :muscle:  :muscle:
Your project meets most of the requirements :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [ ] Don't overuse rubocop disabling
- [ ] #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
      `['dog', 'cat'].any?(/d/) == ['dog', 'cat'].my_any?(/d/)`
- [ ] #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
      `['dog', 'cat'].none?(/d/) == ['dog', 'cat'].my_none?(/d/)`
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by teekaytech — APPROVED** [review]
### Status: Approved :green_circle: 

* * * 

**Congratulations team :tada: 
This project has passed all the mandatory requirements. It is therefore approved :heavy_check_mark:**


Do not forget to fill the project completion form. 
You can now merge to the default branch. :100: 
Good luck with your next project.

**I'm @taofeekolalere on slack.
Happy Coding :man_technologist:**

**Pull request merged**